scylla/src/transport/connection.rs

```diff
@@ -1064,6 +1138,48 @@ impl Connection {
         Ok(())
     }
 
+    // This future enforces driver-side timeouts. It wakes as soon as some
+    // requests expire due to the timeout and sends a timeout error
+    // to the corresponding response handlers.
+    //
+    // The task sometimes wakes spuriously, but infrequently.
+    // See the comments in DeadlineTracker for more information.
+    async fn timeouter(handler_map: &StdMutex<ResponseHandlerMap>) -> Result<(), QueryError> {
+        // We are using poll_fn here in order to be able to poll the expiry timer
+        // inside the deadline tracker without having to hold the mutex.
+        futures::future::poll_fn(|cx| -> Poll<Result<(), QueryError>> {
+            // Guaranteed that nobody else locks the handler_map at the moment
+            let mut handler_map_guard = handler_map.try_lock().unwrap();
+
+            // Wait until the deadline of the first request in the queue elapses
+            futures::ready!(handler_map_guard.deadline_tracker.poll_expiry_timer(cx));
+
+            // Elapse some of the requests
+            let to_elapse = handler_map_guard
+                .deadline_tracker
+                .elapse_some(Instant::now());
+            tracing::debug!("Timing out {} requests", to_elapse.len());
+
+            for request_id in to_elapse {
+                if let Some(handler) = handler_map_guard.orphan(request_id) {
+                    // Ignore sending error, request was dropped
+                    let _ = handler.response_sender.send(Err(QueryError::RequestTimeout(
+                        "request timed out".to_string(), // TODO: include information about deadline?
+                    )));
+                }
+            }
+
+            // The expiry timer should be unset at the moment, so register
+            // the waker right now by polling again
+            futures::ready!(handler_map_guard.deadline_tracker.poll_expiry_timer(cx));
+
+            // We should not get here, but to be 100% safe, let's just yield
+            cx.waker().wake_by_ref();
+            Poll::Pending
+        })
+        .await
```

Comment:
Good catch. With the second `poll_expiry_timer` I wanted the future to immediately start waiting for the next expiration without having to force the whole task to be polled again - this future runs along three other futures in a `try_join`, so most likely this will cause all of them to be polled.

Side note: the `The expiry timer should be unset at the moment` comment is incorrect because `elapse_some` reschedules the timer. I removed it.

I rewrote the function so that runs an infinite loop and uses `poll_fn` only to poll the expiry timer. Entering an infinite loop without preemptions shouldn't be possible as the `Sleep` future used for waiting makes sure to yield if the task runs out of "budget" (see here: https://docs.rs/tokio/latest/tokio/task/index.html#cooperative-scheduling).